src/rgw/rgw_lua_background.h

```diff
@@ -0,0 +1,81 @@
+#pragma once
+#include "common/dout.h"
+#include "rgw_common.h"
+#include <string>
+#include "rgw_lua_utils.h"
+
+namespace rgw::lua {
+
+//Interval between each execution of the script
+//is set to 5 seconds, this setting is configurable
+//using RGW.setoff from within the lua script
+constexpr const int INIT_SET_OFF = 5;
+
+//Writeable meta table named RGW with mutex protection
+using BackgroundMap = std::unordered_map<std::string, std::string>;
+struct RGWTable : StringMapMetaTable<BackgroundMap,
+  StringMapWriteableNewIndex<BackgroundMap>> {
+    static std::string TableName() {return "RGW";}
+    static std::string Name() {return TableName() + "Meta";}
+    static int IndexClosure(lua_State* L) {
+        const auto mtx = reinterpret_cast<std::mutex*>(lua_touserdata(L, lua_upvalueindex(2)));
+        std::lock_guard l(*mtx);
+        return StringMapMetaTable::IndexClosure(L);
+    }
+    static int LenClosure(lua_State* L) {
+        static const auto mtx = reinterpret_cast<std::mutex*>(lua_touserdata(L, lua_upvalueindex(2)));
+        std::lock_guard l(*mtx);
+        return StringMapMetaTable::LenClosure(L);
+    }
+    static int NewIndexClosure(lua_State* L) {
+        static const auto mtx = reinterpret_cast<std::mutex*>(lua_touserdata(L, lua_upvalueindex(2)));
+        std::lock_guard l(*mtx);
+        return StringMapMetaTable::NewIndexClosure(L);
+    }
+};
+
+class Background {
+
+private:
+  BackgroundMap rgw_map;
+  std::string rgw_script;
+  bool stopped = false;
+  int set_off =  INIT_SET_OFF;
+  const DoutPrefixProvider* dpp;
+  rgw::sal::Store* store;
+  std::thread runner;
+  CephContext* cct;
+  lua_State* const L;
+  rgw::lua::lua_state_guard lguard;
+  std::string luarocks_path;
+  std::mutex m_mutex;
+
+  void run();
+
+public:
+  Background(const DoutPrefixProvider* dpp,
+      rgw::sal::Store* const store,
+      CephContext* cct,
+      const std::string& luarocks_path) :
```

Comment:
it seems to me that the use of `RGW.setoff` as the communication channel between the background script and its runner thread is the part that requires this coupling

maybe it makes sense to create a separate `Background` table for `setoff` that's unique to each background script?

Reply:
> i'm having trouble understanding the connection between this global `RGWTable` and the 'background' script/thread. why is the latter necessary for the former?
> 
> i thought the global table was intended for scripts to store information between invocations, or to share data between different scripts
> 
> background scripts sound useful on their own, but the current design looks like it can only support a single background script. and if we have no background script configured, that thread will still wake up and try to read one every 5 seconds?
> 
> can we find a way to decouple these two separate features? ideally we'd be able to create several different tables, and run several different background scripts - and if there are multiple background scripts, we should be able to service all of them from a single thread

Agree that the shared table and the background script are different feature.
The only reason they are here together is that there are usecases where the background script is needed together with the shared table. E.g.
* collecting stats from the request context and periodically sending them from the background script
* doing cleanup in the background for values being cached from the other contexts

Reply:
> it seems to me that the use of `RGW.setoff` as the communication channel between the background script and its runner thread is the part that requires this coupling
> 
> maybe it makes sense to create a separate `Background` table for `setoff` that's unique to each background script?

Agree that it should not be part of the RGW table.
We can just descope from this PR and keep as constant